src/osd/OSD.cc

```diff
@@ -7720,38 +7720,24 @@ void OSD::check_osdmap_features()
   }
 }
 
-bool OSD::advance_pg(
+void OSD::advance_pg(
   epoch_t osd_epoch, PG *pg,
   ThreadPool::TPHandle &handle,
   PG::RecoveryCtx *rctx,
-  set<PGRef> *new_pgs,
-  bool no_max)
+  set<PGRef> *new_pgs)
 {
   assert(pg->is_locked());
   epoch_t next_epoch = pg->get_osdmap()->get_epoch() + 1;
   OSDMapRef lastmap = pg->get_osdmap();
 
-  if (lastmap->get_epoch() == osd_epoch)
-    return true;
   assert(lastmap->get_epoch() < osd_epoch);
 
-  epoch_t min_epoch = service.get_min_pg_epoch();
-  epoch_t max;
-  if (min_epoch) {
-    max = min_epoch + cct->_conf->osd_map_max_advance;
-  } else {
-    max = next_epoch + cct->_conf->osd_map_max_advance;
-  }
-
   for (;
-       next_epoch <= osd_epoch && (no_max || next_epoch <= max);
+       next_epoch <= osd_epoch;
        ++next_epoch) {
     OSDMapRef nextmap = service.try_get_map(next_epoch);
     if (!nextmap) {
       dout(20) << __func__ << " missing map " << next_epoch << dendl;
-      // make sure max is bumped up so that we can get past any
-      // gap in maps
-      max = MAX(max, next_epoch + cct->_conf->osd_map_max_advance);
```

Comment:
Okay, it looks like this max comes from http://tracker.ceph.com/issues/7576, so it's entirely about the in-memory osdmap consumption and freshness of PGs.
So I guess maybe it's okay to just do every PG in one go?

Reply:
Right.  When I looked at this more carefully there are only a few cases that can cause a PG to get behind here:

- a wonkified advance_map on code prior to this patch.  in that case, the "do N maps per pass through advance_map" could get out of sync for different PGs so that they weren't all working on the same range of epochs (e.g., because we were doing that, then crashed, getting the ranges out of sync).  this PR erases that possibility, so that the only time we would hit it woudl be the very first startup (before we are marked up) after an upgrade.
- an imported PG.  ceph-objectstore-tool now catches it up to the latest osdmap epoch, so this isn't possible any more.